File: app/src/main/java/com/commit451/gitlab/dialog/AccessDialog.java
package com.commit451.gitlab.dialog;

import android.content.Context;
import android.support.annotation.NonNull;
import android.view.View;
import android.widget.Toast;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import com.afollestad.materialdialogs.Theme;
import com.commit451.gitlab.App;
import com.commit451.gitlab.R;
import com.commit451.gitlab.model.api.Group;
import com.commit451.gitlab.model.api.Member;
import com.commit451.gitlab.rx.CustomSingleObserver;

import java.util.Arrays;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

/**
 * Change a users access level, either for a group or for a project
 */
public class AccessDialog extends MaterialDialog {

    OnAccessChangedListener onAccessChangedListener;
    Listener listener;

    String[] roleNames;
    long projectId = -1;
    Group group;
    Member member;

    public AccessDialog(Context context, Listener accessAppliedListener) {
        this(context, null, null, -1);
        listener = accessAppliedListener;
    }

    public AccessDialog(Context context, Member member, Group group) {
        this(context, member, group, -1);
    }

    public AccessDialog(Context context, Member member, long projectId) {
        this(context, member, null, projectId);
    }

    private AccessDialog(Context context, Member member, Group group, long projectId) {
        super(new MaterialDialog.Builder(context)
                .items((group == null) ? R.array.project_role_names : R.array.group_role_names)
                .itemsCallbackSingleChoice(-1, new ListCallbackSingleChoice() {
                    @Override
                    public boolean onSelection(MaterialDialog materialDialog, View view, int i, CharSequence charSequence) {
                        return true;
                    }
                })
                .theme(Theme.DARK)
                .progress(true, 0) // So we can later show loading progress
                .positiveText(R.string.action_apply)
                .negativeText(R.string.md_cancel_label));
        roleNames = getContext().getResources().getStringArray((group == null)
                ? R.array.project_role_names
                : R.array.group_role_names);
        getActionButton(DialogAction.POSITIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onApply();
            }
        });
        getActionButton(DialogAction.NEGATIVE).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onCancel();
            }
        });
        this.member = member;
        this.group = group;
        this.projectId = projectId;
        if (this.member != null) {
            setSelectedIndex(Arrays.asList(roleNames).indexOf(
                    Member.getAccessLevel(this.member.getAccessLevel())));
        }
    }

    private void changeAccess(int accessLevel) {

        if (group != null) {
            showLoading();
            editGroupOrProjectMember(App.get().getGitLab().editGroupMember(group.getId(), member.getId(), accessLevel));
        } else if (projectId != -1) {
            showLoading();
            editGroupOrProjectMember(App.get().getGitLab().editProjectMember(projectId, member.getId(), accessLevel));
        } else if (listener != null) {
            listener.onAccessApplied(accessLevel);
        } else {
            throw new IllegalStateException("Not sure what to apply this access change to. Check the constructors plz");
        }
    }

    private void editGroupOrProjectMember(Single<Member> observable) {
        observable
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new CustomSingleObserver<Member>() {

                    @Override
                    public void error(@NonNull Throwable t) {
                        Timber.e(t);
                        AccessDialog.this.onError();
                    }

                    @Override
                    public void success(@NonNull Member member) {
                        if (onAccessChangedListener != null) {
                            onAccessChangedListener.onAccessChanged(AccessDialog.this.member, roleNames[getSelectedIndex()]);
                        }
                        dismiss();
                    }
                });
    }

    public void showLoading() {
        getActionButton(DialogAction.POSITIVE).setEnabled(false);
    }

    private void onError() {
        Toast.makeText(getContext(), R.string.failed_to_apply_access_level, Toast.LENGTH_SHORT).show();
        dismiss();
    }

    public void setOnAccessChangedListener(OnAccessChangedListener listener) {
        onAccessChangedListener = listener;
    }

    private void onApply() {
        if (getSelectedIndex() == -1) {
            Toast.makeText(getContext(), R.string.please_select_access_level, Toast.LENGTH_LONG)
                    .show();
            return;
        }
        String accessLevel = roleNames[getSelectedIndex()];
        if (accessLevel == null) {
            Toast.makeText(getContext(), R.string.please_select_access_level, Toast.LENGTH_LONG)
                    .show();
        } else {
            changeAccess(Member.getAccessLevel(accessLevel));
        }
    }

    private void onCancel() {
        dismiss();
    }

    public interface OnAccessChangedListener {
        void onAccessChanged(Member member, String accessLevel);
    }

    public interface Listener {
        void onAccessApplied(int accessLevel);
    }
}

File: app/src/main/java/com/commit451/gitlab/dialog/AccessDialog.kt
package com.commit451.gitlab.dialog

import android.content.Context
import android.widget.Toast
import com.afollestad.materialdialogs.DialogAction
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.Theme
import com.commit451.gitlab.App
import com.commit451.gitlab.R
import com.commit451.gitlab.model.api.Group
import com.commit451.gitlab.model.api.User
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import io.reactivex.schedulers.Schedulers
import timber.log.Timber

/**
 * Change a users access level, either for a group or for a project
 */
class AccessDialog private constructor(context: Context, internal var member: User?, internal var group: Group?, projectId: Long) : MaterialDialog(Builder(context)
        .items(if (group == null) R.array.project_role_names else R.array.group_role_names)
        .itemsCallbackSingleChoice(-1) { _, _, _, _ -> true }
        .theme(Theme.DARK)
        .progress(true, 0) // So we can later show loading progress
        .positiveText(R.string.action_apply)
        .negativeText(R.string.cancel)) {

    private var onAccessChangedListener: OnAccessChangedListener? = null
    var listener: Listener? = null

    var roleNames: Array<String> = getContext().resources.getStringArray(if (group == null)
        R.array.project_role_names
    else
        R.array.group_role_names)
    var projectId: Long = -1
    val disposables = CompositeDisposable()

    constructor(context: Context, accessAppliedListener: Listener) : this(context, null, null, -1) {
        listener = accessAppliedListener
    }

    constructor(context: Context, member: User, group: Group) : this(context, member, group, -1)

    constructor(context: Context, member: User, projectId: Long) : this(context, member, null, projectId)

    init {
        getActionButton(DialogAction.POSITIVE).setOnClickListener { onApply() }
        getActionButton(DialogAction.NEGATIVE).setOnClickListener { onCancel() }
        this.projectId = projectId
        if (this.member != null) {
            selectedIndex = listOf(*roleNames).indexOf(
                    User.getAccessLevel(this.member!!.accessLevel))
        }
    }

    private fun changeAccess(accessLevel: Int) {
        when {
            group != null -> {
                showLoading()
                editGroupOrProjectMember(App.get().gitLab.editGroupMember(group!!.id, member!!.id, accessLevel))
            }
            projectId != -1L -> {
                showLoading()
                editGroupOrProjectMember(App.get().gitLab.editProjectMember(projectId, member!!.id, accessLevel))
            }
            listener != null -> {
                listener!!.onAccessApplied(accessLevel)
            }
            else -> {
                throw IllegalStateException("Not sure what to apply this access change to. Check the constructors plz")
            }
        }
    }

    private fun editGroupOrProjectMember(observable: Single<User>) {
        disposables += observable
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    if (onAccessChangedListener != null) {
                        onAccessChangedListener!!.onAccessChanged(this@AccessDialog.member!!, roleNames[selectedIndex])
                    }
                    dismiss()
                }, {
                    Timber.e(it)
                    this@AccessDialog.onError()
                })
    }

    fun showLoading() {
        getActionButton(DialogAction.POSITIVE).isEnabled = false
    }

    private fun onError() {
        Toast.makeText(context, R.string.failed_to_apply_access_level, Toast.LENGTH_SHORT).show()
        dismiss()
    }

    fun setOnAccessChangedListener(listener: OnAccessChangedListener) {
        onAccessChangedListener = listener
    }

    private fun onApply() {
        if (selectedIndex == -1) {
            Toast.makeText(context, R.string.please_select_access_level, Toast.LENGTH_LONG)
                    .show()
            return
        }
        val accessLevel = roleNames[selectedIndex]
        changeAccess(User.getAccessLevel(accessLevel))
    }

    private fun onCancel() {
        dismiss()
    }

    override fun onDetachedFromWindow() {
        disposables.clear()
        super.onDetachedFromWindow()
    }

    interface OnAccessChangedListener {
        fun onAccessChanged(member: User, accessLevel: String)
    }

    interface Listener {
        fun onAccessApplied(accessLevel: Int)
    }
}
